### ISSUE TITLE
Route Parameter

### ISSUE THREAD

**Issue by abdollahkahne** [PacktPublishing/Modern-Web-Development-with-ASP.NET-Core-3-Second-Edition#8]
Hi,
Thanks to your advanced book, I am reading it now. I appreciate it if someone response me.
I see this type of route pattern in chapter 3. What does the parenthesis means here? Does it have specific meanings in route templates?
`Calculator/Calculate({a:evenint},{b:evenint})`

If it is a typo then how we can force this constraint in pattern for query strings?! 
Thanks in advance

**Comment by rjperes** [comment]
Hi! Not a typo, this is actually to force its use on the URL, as in /Calculator/Calculate(1,3)

**Comment by abdollahkahne** [comment]
Thanks for your response. 
What a magic! Although I think this does not work for me in Ubuntu (and Kestrel).

**Comment by rjperes** [comment]
I'm pretty sure you're doing something wrong, it does work, regardless of the operating system!
All you need is something like this minimal example:
`[HttpGet("Calculate({a:int},{b:int})")]
        public IActionResult Calculate(int a, int b)
        {
            return Json(a + b);
        }`

**Comment by abdollahkahne** [comment]
Thank you again. Yes it works for me now with attribute routing only. But it not works if I define it in `startup` class.

More Description:
I defined the route pattern only at Startup `Configure` method and it give me the error that the request matched multiple endpoints. After I deleted all the endpoints except the intended action (Just one action in controller) it return the result . This is how I defined endpoint routing in startup and it does not work:
```
endpoints.MapControllerRoute(name:"calculator", 
                pattern:"_api/{controller}/{action}({a:int},{b:int})",
                defaults:new {controller="Calculator",action = "Calculate"},
                constraints:new {controller= new StringRouteConstraint("Calculator")},
                dataTokens:new {Foo ="Bar"}
                );
```
But after changing the above route definition to simple pattern (` pattern:"_api/{controller}/{action}`) and adding route template by attribute to action (Like your comment instruction) it works. 
Although it may not make sense to define that in Startup but I want to add a list of two operand operators in Calculator and the **define a single route template for all of them in Startup `Configure` method**. Should it works too? Or I have to add the route to all methods by `httpxxx` or `route` attributes?
